### PR TITLE
Gitlab storage

### PR DESCRIPTION
**Pull request by apatrushev**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -37,7 +37,7 @@
     },
     install_requires=[
         'aiohttp-graphql',
-        'aiohttp<3.6',
+        'aiohttp!=3.6.0',
```
Comment (mjecov) [inline setup.py:40]: Why `!=` ? Are there any problems with this version?

**Review by mjecov — COMMENTED** [review]
We also have a failing test in tests/test_server.py

File: src/lektorium/app.py [hunk 2]
```diff
@@ -26,32 +35,59 @@
...
+        storage_config = storage_config or 'FILE'
+        storage_type, _, storage_path = storage_config.partition('=')
+        storage_class = StorageType.get(storage_type).value
+        if not storage_path:
+            storage_path = pathlib.Path(closer(tempfile.TemporaryDirectory()))
+            storage_path = storage_class.init(storage_path)
+        storage = storage_class(storage_path)
+
+        lektorium_repo = repo.LocalRepo(storage, server, LocalLektor)
+
```
Comment (dmitriiminaev) [inline src/lektorium/app.py:86]: No `else` condition. If `repo_type` of other value, `lektorium_repo` will not be initialized. Suggest adding `else` with error raising

File: src/lektorium/repo/local/objects.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+class Site(collections.abc.Mapping):
+    ATTR_MAPPING = bidict.bidict({
+        'name': 'site_name',
+        'email': 'custodian_email',
+        'owner': 'custodian',
+    })
+
+    def __init__(self, site_id, production_url, **props):
+        self.data = dict(props)
+        restricted_keys = ('sessions', 'staging_url')
```
Comment (dmitriiminaev) [inline src/lektorium/repo/local/objects.py]: Can be class const

File: src/lektorium/repo/local/objects.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        if set(self.data.keys()).intersection(restricted_keys):
+            raise RuntimeError()
```
Comment (dmitriiminaev) [inline src/lektorium/repo/local/objects.py]: Error could be more descriptive

File: src/lektorium/repo/local/objects.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        self.production_url = production_url
+        self.sessions = {}
+
+    def __getitem__(self, key):
+        if key == 'sessions':
+            return list(self.sessions.values())
+        elif key == 'production_url':
+            result = self.production_url
+            if callable(self.production_url):
+                self.production_url, result = result()
```
Comment (dmitriiminaev) [inline src/lektorium/repo/local/objects.py]: Seems strange that `callable` is checking `production_url` but a call is issued on `result`

File: src/lektorium/repo/local/repo.py [hunk 6]
```diff
@@ -204,11 +112,16 @@ def unpark_session(self, session_id):
 
     def create_site(self, site_id, name, owner=None):
         owner, email = owner or self.DEFAULT_USER
-        master_path = self.root_dir / site_id / 'master'
-        self.lektor.create_site(name, owner, master_path)
+        result = self.storage.create_site(self.lektor, name, owner, site_id)
+        site_root, site_options = result
```
Comment (dmitriiminaev) [inline src/lektorium/repo/local/repo.py]: Why the need for intermediate variable `result`?

File: src/lektorium/repo/local/storage.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
+import abc
+import collections
+import functools
+import inifile
+import pathlib
+import shutil
+import subprocess
+import tempfile
+import yaml
+from .objects import Site
```
Comment (dmitriiminaev) [inline src/lektorium/repo/local/storage.py:11]:
```suggestion

from .objects import Site
```

File: src/lektorium/repo/local/storage.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        lektorium_workdir.mkdir()
+        subprocess.check_call(
+            f'git clone {self.git} .',
+            shell=True,
+            cwd=lektorium_workdir
+        )
+        super().__init__(lektorium_workdir)
+
+    @classmethod
+    def init(cls, path):
```
Comment (dmitriiminaev) [inline src/lektorium/repo/local/storage.py]: Why is this `classmethod` but not `staticmethod`?

File: setup.py (hunk 1 above)
Comment (apatrushev) [reply setup.py:40]: Yes, it is.

File: src/lektorium/repo/local/repo.py (hunk 6 above)
Comment (apatrushev) [reply src/lektorium/repo/local/repo.py]: Only readabilty. I did not find better code shape. I will remove it.

File: src/lektorium/repo/local/storage.py (hunk 8 above)
Comment (apatrushev) [reply src/lektorium/repo/local/storage.py]: The idea was that each variant of storage implementation should have its own init. You are right - at the moment it has no sense.

**Pull request merged**